mailbox/api/src/test/java/org/apache/james/mailbox/events/EventDeadLettersContract.java

```diff
@@ -0,0 +1,184 @@
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.events;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import org.apache.james.core.User;
+import org.apache.james.mailbox.Event;
+import org.apache.james.mailbox.MailboxListener;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.model.MailboxConstants;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.mailbox.model.TestId;
+import org.junit.jupiter.api.Test;
+
+interface EventDeadLettersContract {
+
+    MailboxPath MAILBOX_PATH = new MailboxPath(MailboxConstants.USER_NAMESPACE, "user", "mailboxName");
+    User USER = User.fromUsername("user");
+    MailboxSession.SessionId SESSION_ID = MailboxSession.SessionId.of(235);
+    TestId MAILBOX_ID = TestId.of(563);
+    MailboxListener.MailboxAdded EVENT_1 = new MailboxListener.MailboxAdded(
+        SESSION_ID,
+        USER,
+        MAILBOX_PATH,
+        MAILBOX_ID,
+        Event.EventId.of("6e0dd59d-660e-4d9b-b22f-0354479f47b4"));
+    MailboxListener.MailboxAdded EVENT_2 = new MailboxListener.MailboxAdded(
+        SESSION_ID,
+        USER,
+        MAILBOX_PATH,
+        MAILBOX_ID,
+        Event.EventId.of("6e0dd59d-660e-4d9b-b22f-0354479f47b5"));
+    MailboxListener.MailboxAdded EVENT_3 = new MailboxListener.MailboxAdded(
+        SESSION_ID,
+        USER,
+        MAILBOX_PATH,
+        MAILBOX_ID,
+        Event.EventId.of("6e0dd59d-660e-4d9b-b22f-0354479f47b6"));
+
+    Group GROUP_A = new EventBusContract.GroupA();
+    Group GROUP_B = new EventBusContract.GroupB();
+    Group NULL_GROUP = null;
+    Event NULL_EVENT = null;
+
+    EventDeadLetters eventDeadLetters();
+
+    @Test
+    default void groupsShouldReturnAllStoredGroups() {
+        EventDeadLetters eventDeadLetters = eventDeadLetters();
+        eventDeadLetters.store(GROUP_A, EVENT_1).block();
+        eventDeadLetters.store(GROUP_B, EVENT_1).block();
+
+        assertThat(eventDeadLetters.groups().toStream())
+            .containsExactlyInAnyOrder(GROUP_A, GROUP_B);
```

Comment:
Is order important?

If so what defines the order? Alphabet? First event insertion?

Maybe it is easier to not enforce the order now...

Reply:
according [this](https://stackoverflow.com/questions/47992139/assertj-what-is-the-difference-between-containsonly-and-containsexactlyinanyord) `containsOnly` doesn't care about duplication, but `containsExactlyInAnyOrder` does. In this case, can use both